Inspect qemu monitor to verify rbd is used

junousi opened the pull request:
When iterating qemu PIDs, verify from the qemu monitor 'info block'
section that rbd driver is really being used for any of the VM disks.

Commit:
Inspect qemu monitor to verify rbd is used

When iterating qemu PIDs, verify from the qemu monitor 'info block'
section that rbd driver is really being used for any of the VM disks.

Commit:
Add commented debug

Add commented out section that can be easily uncommented for
getting information about configured RBDs.

oscarkraemer commented on files/ceph-version-qemu.sh:
```diff
@@ -30,11 +30,15 @@ PIDS=$(pidof /usr/libexec/qemu-kvm)
   QEMU_TIMESTAMP=$(date -d "$(stat -c %x /proc/$PID/stat)" +%s)
-  if [[ "$QEMU_TIMESTAMP" -lt "$CEPH_UPGRADE_TIMESTAMP" ]]
+  INSTANCE_NAME=$(ps -ef -q $PID|grep -Po "(instance-\w+)"|uniq)
+  INSTANCE_QEMU_MON_RBD=$(virsh qemu-monitor-command --hmp "$INSTANCE_NAME" 'info block'|grep rbd)
```
What does this do? INSTANCE_QEMU_MON_RBD

oscarkraemer replied:
Don't be afraid to comment your code.

junousi replied:
I got rid of my fears and commented

oscarkraemer commented on files/ceph-version-qemu.sh:
```diff
@@ -30,11 +30,15 @@ PIDS=$(pidof /usr/libexec/qemu-kvm)
-    INSTANCE_NAME=$(ps -ef -q $PID|grep -Po "(instance-\w+)"|uniq)
     INSTANCE_UUID=$(virsh dumpxml $INSTANCE_NAME|grep \/uuid|awk -F'>' "{print \$2}"|awk -F'<' "{print \$1}")
     echo $INSTANCE_UUID
+    #INSTANCE_QEMU_MON_RBD_COUNT=$(virsh qemu-monitor-command --hmp "$INSTANCE_NAME" 'info block'|grep -c rbd)
```
I think you need to add a better comment here.

junousi replied:
Added

oscarkraemer commented on files/ceph-version-qemu.sh:
```diff
@@ -30,11 +30,15 @@ PIDS=$(pidof /usr/libexec/qemu-kvm)
+  INSTANCE_NAME=$(ps -ef -q $PID|grep -Po "(instance-\w+)"|uniq)
+  INSTANCE_QEMU_MON_RBD=$(virsh qemu-monitor-command --hmp "$INSTANCE_NAME" 'info block'|grep rbd)
+  INSTANCE_USE_RBD=$?
+  if [[ "$QEMU_TIMESTAMP" -lt "$CEPH_UPGRADE_TIMESTAMP" ]] && [ "$INSTANCE_USE_RBD" -eq "0" ]
```
How does this works if an instance has been booted on the same day as ceph was upgraded?

junousi replied:
Yum event begin time is the reference point, anything before that is considered to be the old version

Commit:
Add comments

Comment most things in the script. Also add more exhaustive
command for retrieving yum history event IDs.

Commit:
Use proper options and add some

Move previous arguments to getopt for processing.

Add a flag for excluding Qemus that don't have any RBDs. An example
of this is would be e.g. an io.* VM which was booted in Jewel era
and does not have any volumes attached.

Add a debug flag which allows outputting individual RBD counts (since
that seems to correlate nicely to the number that `ceph --features`
outputs.

Commit:
Fix getopt params

Also make debug configurable from role variable

oscarkraemer commented on files/ceph-version-qemu.sh:
```diff
@@ -3,38 +3,82 @@
 # By default, search for upgrades from Jewel to Luminous.
 CEPH_MAJORVER_EXPECTED_OLD=10
 CEPH_MAJORVER_EXPECTED_NEW=12
+# By default, exclude VMs that do not have RBD block devices attached.
```
This does not reflect my test. When I use -e it returns two instances that does not have a rbd volume, while when not using -e it returns nothing even if one instances has an rbd volume.

junousi replied:
Too much caffeine when writing, the logic was backwards. Should be fixed now.

oscarkraemer commented on files/ceph-version-qemu.sh:
```diff
@@ -3,38 +3,82 @@
-if [ -z ${CEPH_MAJORVER_EXPECTED_NEW+x} ]; then
-  CEPH_MAJORVER_EXPECTED_NEW=$2
-fi
+usage() { echo "Usage: $0 [-o 10] [-n 12] [-e] [-d]" 1>&2; exit 1; }
```
Probably better documentation would be welcomed.

junousi replied:
Added

Commit:
Fix logic

Things make more sense if we by default include rbd-less qemus
and change the optional parameter to exclude.

Commit:
Add debug and exclude as proper ansible variables

Commit:
Fix variable naming

Commit:
Change include into exclude

Using an exclude argument to control an include variable was a
bit convoluted, so sort that out. Also add a "Managed by Ansible"
header.

martbhell commented on vars/main.yml:
```diff
@@ -1,2 +1,4 @@
\ No newline at end of file
+# vars file for ansible-role-qemu-ceph-clients
+script_debug: false
+exclude_no_rbd: false
```
vars/ directory has quite much [1] higher priority than the defaults/ directory

To increase usability of the role they could be put in defaults/

script_debug is a quite generic name but I can't find it re-used in any other roles so probably fine.

exclude_no_rbd I couldn't figure out what it meant by reading the variable but had to read the bash script. Perhaps it could be changed to something more descriptive like "exclude_qemus_without_rbds" or some such?

[1]: https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable

junousi replied:
An unfortunate case of "I don't frequent this bar that often." Moved them to defaults, also renamed.

martbhell replied:
I can't run with only "-e script_debug_ouput=True", have to also use the -e debug. 
Perhaps that's a feature?
Setting DEBUG=0 when script_debug_output=True caught me off card (and as a side bonus thing caused me to read the script a bit in more detail). If it's just me then keep as is, but somehow feels better that DEBUG=1 and EXCLUDE_=1 when debug is enabled and when we are excluding.

junousi replied:
Looks like the "empty" option in the ternary doesn't work as it leaves a trail of double quotes

```
# ./ceph-version-qemu.sh "-d"
UUID
DEBUG: UUID has this many RBD devices configured: 0
```

works

```
# ./ceph-version-qemu.sh "" "-d"
UUID
UUID
```

doesn't.

I am starting to strongly dislike this task. :D

But stay tuned for a fix.

junousi replied:
Fixed

martbhell commented on tasks/main.yml:
```diff
@@ -9,5 +9,5 @@
   no_log: true
 - name: Run verification script
-  shell: /root/ceph-version-qemu.sh
+  shell: /root/ceph-version-qemu.sh "{{ '-e' if exclude_no_rbd=='true' else '' }}" "{{ '-d' if script_debug=='true' else '' }}"
```
I can't get the extra arguments to work with ansible.

```
$ ansible-playbook -i inventories/hosts.file ops_playbooks/ceph-version-qemu.yml -e script_debug=True  -vvv
```

Does not print the debug output from the bash script:

```
ok: [mycloudnode2.example.org] => {
    "changed": false, 
    "cmd": "/root/ceph-version-qemu.sh \"\" \"\"", 
    "delta": "0:00:02.789636", 
    "end": "2019-03-14 12:36:31.198899", 
    "invocation": {
        "module_args": {
            "_raw_params": "/root/ceph-version-qemu.sh \"\" \"\"", 
            "_uses_shell": true, 
            "chdir": null, 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "stdin": null, 
            "warn": true
        }
    }, 
    "rc": 0, 
    "start": "2019-03-14 12:36:28.409263", 
    "stderr": "", 
    "stderr_lines": [], 
    "stdout": "UID UID2, 
    "stdout_lines": [
        "UID2", 
        "UID1d"
    ]
}
```

I think that's because the script is asking for true, which apparently is not a boolean. Maybe that could be adjusted? (then also putting in True / False or yes/no and the other yaml variants would work)

junousi replied:
That's a clear bug that in my testing happened to work since I used true and it was expecting a literal 'true'. Should be fixed now

Commit:
Move variables from vars to default

Also rename them to something hopefully better. Also fix an
error confusing a literal 'true' to boolean true.

Commit:
Fix a quoting issue

Instead of quoting individual "{{statements}}" quote the whole
command.
